src/System.Private.CoreLib/shared/System/Collections/Generic/List.cs

```diff
@@ -165,6 +165,8 @@ public int Capacity
             }
         }
 
+        internal Span<T> AsSpan() => new Span<T>(_items, 0, _size);
```

Comment:
To be "safe" and match what other mutating APIs do, `AsSpan()` needs to increment the version count.

`AsReadOnlySpan` does not have this requirement.

Reply:
Is it really needed to increment the version stamp? This is unsafe API. You have to know what you are doing.

Reply:
Is there a way for users to ensure the version count is incremented without us doing it here?

The assumption is that, even if you are using this API for perf benefits, you still want to do the "correct" thing with regards to other usages (such as fulfilling the `IEnumerable<T>` contract `List<T>` is currently exposing).

Its a trivial operation to increment the version count once and makes it more correct. For the case where you aren't mutating you can just get the `ReadOnlySpan<T>`. If you really want to silently mutate and avoid the version increment, you can then use `MemoryMarshal` to get a writeable span from the `ROSpan`.

Reply:
Doesn't mimic List's behaviour; it should increment the version count on each change rather than just obtaining the Span window (which then may not change anything). That can't be done; also it would defeat the purpose of the Api.

That said if you are using the Span and enumerating the List at the same time; would be an unfortunate approach as enumerating over the Span will be much faster (as it can remove the bounds checks and  skip the enumerator creation) and you already have that.

Reply:
> Doesn't mimic List's behaviour; it should increment the version count on each change rather than just obtaining the Span window (which then may not change anything).

Fair enough. The only downside is that it seems not ideal that users can't easily ensure "correct" versioning even if using the `Span<T>` APIs